tests/README.md

```diff
@@ -0,0 +1,1510 @@
+
+To summarize, it is very important to understand the data flow in the system to reduce unnecessary unit tests.
+
+<a name="ordering"></a>
+#### Ordering
+The order of data or the position of specific data in larger collections are often a point were something
+goes wrong within a system.
+
+<a name="range"></a>
+#### Range
+The the 64bit integer of PHP has far more capacity than needed. For example the age of a dog will never exceed a certain point, however if something went wrong the dog gets 2 pow 64 years old.
+
+The excessive usage of primitives is known as a code smell with the name *primitive obsession*. One of the primal benefit of PHP is that data can be abstracted with its own logic. For example a dog has at most four legs and its age is between 1 second and 30 years.  
+
+To abstract these values and test the constraints of the abstraction makes the rest of the application more resistant again such errors.
+
+<a name="reference"></a>
+#### Reference
+When a method is tested the following criteria should be considered:
+- What is the method referencing outside of the scope
+- Which dependencies are there
+- If the method depends on objects being in a specific state
+- Other conditions which must exist for the method
+
+If assumptions are made about a state, the code should be tested that it is not behaving in a wrong way
+when the assumption is not true. For example a plane has to expand the wheels before landing or the plane will most likely be destroyed
+after the landing. Therefore, a plane must transition into the right state before doing a certain action. This situations must be
+tested or it is almost certain that something goes wrong in the future.
+
+```php
+<?php
+/**
+ * @Test
+ */
+public function testPlaneLandingWithExpandedWheelsWhichShouldSucceed() {
+
+	//arrange (preconditions)
+	$plane = new Plane();
+	$plane->start();
+	$plane->expandWheels();
+
+	//act
+	$plane->land();
+
+	//assert (postconditions)
+	$this->assertSame(0, $plane->getSpeed());
+}
+```
+
+A fun fact of this unit test, if the *expandWheels* method is not behaving well in case of failure
+and just ignores the fact that the wheels can't be used. The plane would crash and the speed would also equals zero. Therefore, a green unit test is displayed, the plane landed but not as expected! However if the class is fully tested some or at least one of the *expandWheels* tests will fail and the error is easily spotted and fixed.
+
+<a name="existence"></a>
+#### Existence
+A potentially large sum of defects could be discovered by asking the question "Does something given exist?" For a method which has parameters
+or accessing fields should be thought about if they can be null and how the code should behave in such a case.
+
+Sadly, at the time a null value gets into the wrong place it is often not easy to tell where the actual problems are. Using exceptions
+to tell the consumer of code what went wrong greatly simplifies the search after a problem.
```

Comment:
Yes if a value can be absent, it should be tested that the code behaves like expected and throws an exception if the value must be present. I think it is important that the exceptions actually tell what the problem is and that the function call never returns null in case of a failure.